pwiz_tools/Shared/PanoramaClient/RequestHelper.cs

```diff
@@ -0,0 +1,317 @@
+﻿using System;
+using System.Collections.Specialized;
+using System.Net;
+using System.Text;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using pwiz.PanoramaClient.Properties;
+
+namespace pwiz.PanoramaClient
+{
+    public interface IRequestHelper : IDisposable
+    {
+        JObject Get(Uri uri, string messageOnError = null);
+        JObject Post(Uri uri, NameValueCollection postData, string messageOnError = null);
+        JObject Post(Uri uri, string postData, string messageOnError);
+        void DoRequest(HttpWebRequest request, string method, string authHeader, string messageOnError = null);
+        void RequestJsonResponse();
+        void AddHeader(string name, string value);
+        void RemoveHeader(string name);
+        void AsyncUploadFile(Uri address, string method, string fileName);
+        void CancelAsyncUpload();
+        void AddUploadFileCompletedEventHandler(UploadFileCompletedEventHandler handler);
+        void AddUploadProgressChangedEventHandler(UploadProgressChangedEventHandler handler);
+    }
+
+    public abstract class AbstractRequestHelper : IRequestHelper
+    {
+        private const string APPLICATION_JSON = @"application/json";
+
+
+        #region IRequestHelper methods
+
+        public abstract void AddHeader(string name, string value);
+
+        public abstract void RemoveHeader(string name);
+
+        public abstract void CancelAsyncUpload();
+
+        public abstract void AddUploadFileCompletedEventHandler(UploadFileCompletedEventHandler handler);
+
+        public abstract void AddUploadProgressChangedEventHandler(UploadProgressChangedEventHandler handler);
+
+        public abstract void Dispose();
+
+        #endregion
+
+        public abstract string DoGet(Uri uri);
+
+        public abstract byte[] DoPost(Uri uri, NameValueCollection postData);
+
+        public abstract string DoPost(Uri uri, string postData); // Used only in AuditLogTutorialTest
+
+        public abstract void DoAsyncFileUpload(Uri address, string method, string fileName);
+        
+        public abstract void AddHeader(HttpRequestHeader header, string value);
+
+        public abstract string GetResponse(HttpWebRequest request);
+
+        public abstract Func<WebException, LabKeyError> GetWebExceptionResponseReader();
```

Comment:
It's unusual to have an abstract method which returns a Func.
Usually this would be:
`public LabKeyError GetErrorFromWebException(WebException webException)`
Then, if you needed a Func to pass to "GetErrorFromWebException" itself, because if you refer to a method name without any parentheses you can use it the way you would a Func or an Action.